### PR TITLE
feat: Add support for AppMesh controller addon with AppMesh mTLS example

### PR DESCRIPTION
**Pull request by svelango**
### What does this PR do?

Adds AppMesh Addon

### Motivation

Closes tickets as much as possible

### More

- [ X] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [x] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [x] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [X ] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +65,7 @@ module "eks-blueprints-kubernetes-addons" {
     enable_aws_for_fluentbit              = true
     enable_argocd                         = true
     enable_ingress_nginx                  = true
+    enable_appmesh                        = true
```
Comment (kcoleman731) [inline README.md:68]: For consistency, let's align with the name of the chart which is `appmesh-controller`. So this should be `enable_appmesh_controller` and the module dir should be `appmesh-controller` as opposed to `aws-app-mesh

File: modules/kubernetes-addons/aws-app-mesh/locals.tf [hunk 2]
```diff
@@ -0,0 +1,49 @@
+locals {
+  namespace            = "appmesh-system"
+  name                 = "aws-app-mesh"
```
Comment (kcoleman731) [inline modules/kubernetes-addons/aws-app-mesh/locals.tf:3]: name should be `appmesh-controller`

File: modules/kubernetes-addons/aws-app-mesh/locals.tf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  service_account_name = "${local.name}-sa"
+  aws_region_name      = data.aws_region.current.name
+
+  default_helm_config = {
+    name             = "appmesh-controller"
```
Comment (kcoleman731) [inline modules/kubernetes-addons/aws-app-mesh/locals.tf:8]: update name local and use here.

File: modules/kubernetes-addons/aws-app-mesh/locals.tf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  )
+
+  irsa_config = {
+    kubernetes_namespace              = local.helm_config["namespace"]
+    kubernetes_service_account        = local.service_account_name
+    create_kubernetes_namespace       = true
+    create_kubernetes_service_account = true
+    irsa_iam_policies                 = concat([aws_iam_policy.appmesh.arn], var.irsa_policies)
+  }
+
```
Comment (kcoleman731) [inline modules/kubernetes-addons/aws-app-mesh/locals.tf:33]: extra new line

**Review by kcoleman731 — COMMENTED** [review]
Looks great. Thank you for the PR! Minor comments for you.

**Review by askulkarni2 — COMMENTED** [review]
@svelango thanks for the PR. We will need some additional things:
1. For gitops mode, please open another PR [here](https://github.com/aws-samples/eks-blueprints-add-ons). See other addons for example for what's needed.
2. Add documentation under `docs`
3. Add a detailed example under `examples` to demonstrate a non-trivial usage of the add-on.
4. Provide some information in the PR comments on how this was tested.

File: modules/kubernetes-addons/main.tf [hunk 5]
```diff
@@ -95,6 +95,14 @@ module "aws_node_termination_handler" {
   addon_context           = local.addon_context
 }
 
+module "aws_app_mesh" {
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/main.tf:98]: rename this to 

```hcl
module "appmesh-controller" {}
```

File: modules/kubernetes-addons/main.tf [hunk 6]
```diff
@@ -95,6 +95,14 @@ module "aws_node_termination_handler" {
...
+  count                   = var.enable_appmesh  ? 1 : 0
+  source                  = "./aws-app-mesh"
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/main.tf:100]: rename this folder to `appmesh-controller`

File: modules/kubernetes-addons/variables.tf [hunk 7]
```diff
@@ -113,6 +113,25 @@ variable "cluster_autoscaler_helm_config" {
   description = "Cluster Autoscaler Helm Chart config"
 }
 
+#-----------AWS Appmesh-------------
+variable "enable_appmesh" {
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/variables.tf:117]: change this to `enable_appmesh_controller`

File: modules/kubernetes-addons/aws-app-mesh/locals.tf [hunk 8]
```diff
@@ -0,0 +1,49 @@
+locals {
+  namespace            = "appmesh-system"
+  name                 = "aws-app-mesh"
+  service_account_name = "${local.name}-sa"
```
Comment (vara-bonthu) [inline modules/kubernetes-addons/aws-app-mesh/locals.tf:4]: We can remove the `-sa` at the end service_account_name.

**Review by vara-bonthu — COMMENTED** [review]
Thanks for the PR @svelango  Few minor comments needs incorporating and the comments from @kcoleman731 

- We also need a docs page for this Add-on here https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs/add-ons. Refer to the other Add-ons as an example

- Update the PR with your test results for this add-on.

**Comment by svelango** [comment]
The examples directory is there. Need to see why build failed. Checking.

File: modules/kubernetes-addons/variables.tf [hunk 9]
```diff
@@ -156,6 +156,25 @@ variable "coredns_autoscaler_helm_config" {
   default     = {}
 }
 
+#-----------AWS Appmesh-------------
+variable "enable_appmesh_controller" {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/variables.tf:160]: variable params are in the order of: 
```hcl
variable "foo" {
  description = "A variable"
  type        = string
  default     = null
}
```

File: modules/kubernetes-addons/appmesh-controller/variables.tf [hunk 10]
```diff
@@ -0,0 +1,34 @@
+variable "irsa_policies" {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/appmesh-controller/variables.tf:1]: See comment below regarding variable param ordering

File: modules/kubernetes-addons/appmesh-controller/data.tf [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      "appmesh:DeleteVirtualService",
+      "appmesh:DeleteVirtualNode",
+      "appmesh:DeleteVirtualGateway"
+    ]
+  }
+
+  statement {
+    sid       = "CreateServiceLinkedRole"
+    effect    = "Allow"
+    resources = ["arn:aws:iam::*:role/aws-service-role/appmesh.amazonaws.com/AWSServiceRoleForAppMesh"]
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/appmesh-controller/data.tf:51]: the partition and dns suffix should use the context object to pull for the appropriate region - see https://github.com/terraform-aws-modules/terraform-aws-iam/blob/ed6683e0a3ace4abaa3729e65d56259e09061812/modules/iam-role-for-service-accounts-eks/policies.tf#L957

File: examples/appmesh-mtls/versions.tf [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+      source  = "hashicorp/aws"
+      version = ">= 3.72"
+    }
+    kubectl = {
+      source  = "gavinbunney/kubectl"
+      version = ">= 1.14"
+    }
+  }
+
+  backend "local" {
```
Comment (bryantbiggs) [inline examples/appmesh-mtls/versions.tf:15]: please follow the existing format - see https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/examples/tls-with-aws-pca-issuer/versions.tf#L19-L24

```hcl
  # ##  Used for end-to-end testing on project; update to suit your needs
  # backend "s3" {
  #   bucket = "terraform-ssp-github-actions-state"
  #   region = "us-west-2"
  #   key    = "e2e/appmesh-mtls/terraform.tfstate"
  # }
```

File: examples/appmesh-mtls/variables.tf [hunk 13]
```diff
@@ -0,0 +1,17 @@
+variable "eks_cluster_domain" {
```
Comment (bryantbiggs) [inline examples/appmesh-mtls/variables.tf:1]: See note below on variable param ordering

File: examples/appmesh-mtls/main.tf [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+        "client auth"
+      ]
+      privateKey = {
+        algorithm : "RSA"
+        size : 2048
+      }
+    }
+  })
+
+  depends_on = [
```
Comment (bryantbiggs) [inline examples/appmesh-mtls/main.tf:230]: why does this need to depend on the addons?

File: examples/appmesh-mtls/main.tf [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+      name = module.eks_blueprints.eks_cluster_id
+    }
+
+    spec = {
+      arn = aws_acmpca_certificate_authority.example.arn
+      region : local.region
+    }
+  })
+
+  depends_on = [module.eks_blueprints.managed_node_groups]
```
Comment (bryantbiggs) [inline examples/appmesh-mtls/main.tf:190]: why does this need to depend on the node group?

File: examples/appmesh-mtls/main.tf [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+  azs      = slice(data.aws_availability_zones.available.names, 0, 3)
+
+
+  tags = {
+    Blueprint  = local.name
+    GithubRepo = "github.com/aws-ia/terraform-aws-eks-blueprints"
+  }
+}
+
+#---------------------------------------------------------------
```
Comment (bryantbiggs) [inline examples/appmesh-mtls/main.tf:47]: move this supporting resources/VPC to bottom of example

File: README.md [hunk 17]
```diff
@@ -45,24 +45,30 @@ module "eks_blueprints" {
   }
 }
 
-module "eks_blueprints_kubernetes_addons" {
```
Comment (bryantbiggs) [inline README.md:48]: lets not modify the main `README.md`

**Comment by bryantbiggs** [comment]
@svelango is ACM PCA required for AppMesh mTLS or can folks use cert-manager as the CA? If we can demonstrate this with cert-manager I think that would make the example more applicable to a broader audience and also easier for folks to get up and running with.

Ref: https://github.com/aws/aws-app-mesh-roadmap/issues/34

File: modules/kubernetes-addons/variables.tf (hunk 9 above)
Comment (svelango) [reply modules/kubernetes-addons/variables.tf:160]: Done

File: modules/kubernetes-addons/appmesh-controller/data.tf (hunk 11 above)
Comment (svelango) [reply modules/kubernetes-addons/appmesh-controller/data.tf]: Nice. Learnt something new

File: examples/appmesh-mtls/main.tf (hunk 14 above)
Comment (svelango) [reply examples/appmesh-mtls/main.tf:188]: Removed it

File: examples/appmesh-mtls/main.tf (hunk 16 above)
Comment (svelango) [reply examples/appmesh-mtls/main.tf:47]: Done

File: examples/appmesh-mtls/main.tf (hunk 15 above)
Comment (svelango) [reply examples/appmesh-mtls/main.tf]: Removed

**Comment by svelango** [comment]
> @svelango is ACM PCA required for AppMesh mTLS or can folks use cert-manager as the CA? If we can demonstrate this with cert-manager I think that would make the example more applicable to a broader audience and also easier for folks to get up and running with.
> 
> Ref: [aws/aws-app-mesh-roadmap#34](https://github.com/aws/aws-app-mesh-roadmap/issues/34)

@bryantbiggs Picked ACM as we are adding ACM addon into EKS so as a example added a sample for ACM.

**Comment by bryantbiggs** [comment]
> > @svelango is ACM PCA required for AppMesh mTLS or can folks use cert-manager as the CA? If we can demonstrate this with cert-manager I think that would make the example more applicable to a broader audience and also easier for folks to get up and running with.
> > Ref: [aws/aws-app-mesh-roadmap#34](https://github.com/aws/aws-app-mesh-roadmap/issues/34)
> 
> @bryantbiggs Picked ACM as we are adding ACM addon into EKS so as a example added a sample for ACM.

Lets use cert-manager here please

**Comment by svelango** [comment]
> 



> > > @svelango is ACM PCA required for AppMesh mTLS or can folks use cert-manager as the CA? If we can demonstrate this with cert-manager I think that would make the example more applicable to a broader audience and also easier for folks to get up and running with.
> > > Ref: [aws/aws-app-mesh-roadmap#34](https://github.com/aws/aws-app-mesh-roadmap/issues/34)
> > 
> > 
> > @bryantbiggs Picked ACM as we are adding ACM addon into EKS so as a example added a sample for ACM.
> 
> Lets use cert-manager here please

hi @bryantbiggs If I understand we wanted to use Cert-Manager and ACM PCA as Issuer and not directly work with ACM to get certificate?

File: README.md [hunk 18]
```diff
@@ -45,24 +45,30 @@ module "eks_blueprints" {
...
-  enable_aws_for_fluentbit            = true
-  enable_aws_load_balancer_controller = true
-  enable_cluster_autoscaler           = true
-  enable_metrics_server               = true
-  enable_prometheus                   = true
+#--------------------------------------------
+# Deploy Kubernetes Add-ons with sub module
+#--------------------------------------------
+module "eks-blueprints-kubernetes-addons" {
+    source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/kubernetes-addons"
```
Comment (allamand) [inline README.md:52]: we should specify version here

File: README.md [hunk 19]
```diff
@@ -45,24 +45,30 @@ module "eks_blueprints" {
...
+    #K8s Add-ons
+    enable_aws_load_balancer_controller   = true
+    enable_metrics_server                 = true
+    enable_cluster_autoscaler             = true
+    enable_aws_for_fluentbit              = true
+    enable_argocd                         = true
+    enable_ingress_nginx                  = true
+    enable_appmesh_controller             = true
+
+    depends_on = [module.eks-blueprints.managed_node_groups]
```
Comment (allamand) [inline README.md:71]: depend_on is not recommended with modules

**Review by allamand — CHANGES_REQUESTED** [review]
There is not output from terraform, so how can we connect to the cluster ?

Also It missing some Readme with guidance on how to uses this AppMesh add-on with a simple example to follow

**Review by bryantbiggs — APPROVED** [review]

**Pull request merged**